### PR TITLE
fixed unit tests

### PR DESCRIPTION
**Pull request by nalbion**
Note that there are still some tests failing, and they are likely to have an impact on users.

https://github.com/Pythagora-io/gpt-pilot/actions

They were passing prior to this commit: https://github.com/Pythagora-io/gpt-pilot/commit/6c571f0946a0498ee9b623455034177ee7a4c589

I'm not sure would we want to replace `~` with `''` in `get_full_file_path()`

```python
        file_path = file_path.replace('~', '')
        file_name = file_name.replace('~', '')
```

We probably should have some safe-guards around the LLM reading/writing arbitrary files outside of the workspace folder, but it could be helpful if the the system was able to edit `~/.gitconfig`, `/etc/hosts` etc

Also, these lines are breaking some valid cases in `test_save_file()`

> {'name': 'Dockerfile', 'path': None, 'saved_to': '/temp/gpt-pilot-test/Dockerfile'},

```python
        if '.' not in file_path and not file_path.endswith('/'):
            file_path += '/'
        if '.' not in file_name and not file_name.endswith('/'):
            file_name += '/'
```

> {'name': '/etc/hosts', 'path': None, 'saved_to': '/etc/hosts'},

```python
        if '/' in file_path and not file_path.startswith('/'):
            file_path = '/' + file_path
        if '/' in file_name and not file_name.startswith('/'):
            file_name = '/' + file_name
```

### CODE REVIEW

**Comment by nalbion** [comment]
I'm not sure why `TestCodeMonkey` and `TestDeveloper` are failing in CI - they pass for me on Windows _and_ WSL Ubuntu

File: pilot/utils/files.py [hunk 1]
```diff
@@ -23,7 +31,6 @@ def setup_workspace(args):
         return args['workspace']
 
     root = args.get('root') or get_parent_folder('pilot')
-    create_directory(root, 'workspace')
```
Comment (LeonOstrez) [inline pilot/utils/files.py:26]: why is this removed?

Comment (nalbion) [reply pilot/utils/files.py:26]: Because `os.makedirs` does this anyway

>def makedirs(name, mode=0o777, exist_ok=False):
>    """
>    Super-mkdir; create a leaf directory and all intermediate ones.  Works like
>    mkdir, except that any intermediate path segment (not just the rightmost)
>    will be created if it does not exist.

**Comment by nalbion** [comment]
- added Python 3.12 to test matrix
- resolve #126  - Docker build requires 3.11 (for now)
- removed imports from `pilot/helpers/__init__.py` as it was breaking mocks on Python 3.9 & 3.10
- raise new `ApiKeyNotDefinedError`
- TODO: Treatment of paths outside of the project workspace - https://github.com/Pythagora-io/gpt-pilot/issues/129
- logger now logs at WARN and ERROR
